Bluetooth: Controller: Add coexistence ticker

optical-o opened the pull request:
Implementation of radio coexistence resolves #50463.

cvinayak commented on dts/bindings/net/wireless/generic,coex-gpio.yaml:
```diff
@@ -0,0 +1,19 @@
+description: |
+    Generic representation of Coexistance pinx for radio
```
```suggestion
    Generic representation of Co-Existance pin for radio
```

cvinayak commented on samples/bluetooth/beacon/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright 2019 NXP
```
Update copyright

cvinayak commented on subsys/bluetooth/controller/CMakeLists.txt:
```diff
@@ -155,6 +155,10 @@ if(CONFIG_BT_LL_SW_SPLIT)
     CONFIG_BT_HCI_MESH_EXT
     ll_sw/ll_mesh.c
     )
+  zephyr_library_sources_ifdef(
+    CONFIG_BT_CTLR_COEX_TICKER
+    ll_sw/coex_ticker.c
+    )
```
Place the implementation under `drivers/bluetooth/coex`

```suggestion
```

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -915,6 +915,12 @@ config BT_CTLR_USER_CPR_INTERVAL_MIN
 
 endmenu
 
+config BT_CTLR_COEX_TICKER
+	bool "Coexistence Ticker"
+	help
+	  When enabled Coexistence device implementation is included in
+	  the controller. Which abort any radio states, when coex pin is asserted.
+
```
ditto, Have Kconfig in `drivers/bluetooth/Kconfig`
```suggestion
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/coex_ticker.c:
```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Update copyright

cvinayak commented on subsys/bluetooth/controller/ll_sw/coex_ticker.c:
```diff
@@ -0,0 +1,258 @@
+}
+
+
+#if defined(CONFIG_BT_CTLR_DEBUG_PINS) || \
+	defined(CONFIG_BT_CTLR_DEBUG_PINS_CPUAPP)
+#define DEBUG_PORT NRF_P0
+#define DEBUG_PIN_INHIBIT BIT(23)
+#define DEBUG_PIN_IRQ BIT(25)
+#define DEBUG_PIN_MASK    (DEBUG_PIN_IRQ | DEBUG_PIN_INHIBIT)
+#define DEBUG_INIT() \
+	do { \
+		DEBUG_PORT->DIRSET = DEBUG_PIN_MASK; \
+		DEBUG_PORT->OUTCLR = DEBUG_PIN_MASK; \
+	} while (0)
+
+#define DEBUG_INHIBIT(flag) \
+	do { \
+		if (flag) { \
+			DEBUG_PORT->OUTSET = DEBUG_PIN_INHIBIT; \
+		} else { \
+			DEBUG_PORT->OUTCLR = DEBUG_PIN_INHIBIT; \
+		} \
+	} while (0)
+
+
+#define DEBUG_IRQ(flag) \
+	do { \
+		if (flag) { \
+			DEBUG_PORT->OUTSET = DEBUG_PIN_IRQ; \
+		} else { \
+			DEBUG_PORT->OUTCLR = DEBUG_PIN_IRQ; \
+		} \
+	} while (0)
+#else
+#define DEBUG_INIT()
+#define DEBUG_INHIBIT(flag)
+#define DEBUG_IRQ(flag)
+#endif
+
```
Add the debug macros here (`DEBUG_COEX_INHIBIT` and `DEBUG_COEX_IRQ`): https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/debug.h#L321

```suggestion
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/coex_ticker.c:
```diff
@@ -0,0 +1,258 @@
+#include <logging/log.h>
+#include <drivers/gpio.h>
+#include <sys/__assert.h>
+#include <bluetooth/hci.h>
+#include "controller/hal/ticker.h"
```
blank line to separate the controller includes.

```suggestion
#include <bluetooth/hci.h>

#include "controller/hal/ticker.h"
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/coex_ticker.c:
```diff
@@ -0,0 +1,258 @@
+	struct gpio_callback inhibit_irq_cb;
+};
+
+static inline int _ticker_stop(uint8_t inst_idx, uint8_t u_id, uint8_t tic_id)
+{
+	int ret = ticker_stop(inst_idx, u_id, tic_id, NULL, NULL);
+
+	if (ret != TICKER_STATUS_SUCCESS &&
+		ret != TICKER_STATUS_BUSY) {
+		__ASSERT(0, "Failed to stop ticker.\n");
+	}
+
+	return ret;
+}
```
Use the contents directed in the caller.

```suggestion
static inline int _ticker_stop(uint8_t inst_idx, uint8_t u_id, uint8_t tic_id)
{
	int ret = ticker_stop(inst_idx, u_id, tic_id, NULL, NULL);

	if (ret != TICKER_STATUS_SUCCESS &&
		ret != TICKER_STATUS_BUSY) {
		__ASSERT(0, "Failed to stop ticker.\n");
	}

	return ret;
}
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/coex_ticker.c:
```diff
@@ -0,0 +1,258 @@
+	uint8_t ticker_id;
+	uint32_t err;
+
+	ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
```
Have a new interface in controller and have a unique instance for co-ex, so that the ticker instance does not overlap flash driver use.

cvinayak commented on subsys/bluetooth/controller/ll_sw/coex_ticker.c:
```diff
@@ -0,0 +1,258 @@
+	return ret;
+}
+
+static int time_slot_delay(uint32_t ticks_at_expire, uint32_t ticks_delay,
+				ticker_timeout_func callback, void *context)
+{
+	uint8_t instance_index;
+	uint8_t ticker_id;
+	int err;
+
+	ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
+
+	/* start a secondary one-shot ticker after ticks_delay,
+	 * this will let any radio role to gracefully abort and release the
+	 * Radio h/w.
+	 */
+	err = ticker_start(instance_index, /* Radio instance ticker */
+				1, /* user id for link layer ULL_HIGH */
+				   /* (MAYFLY_CALL_ID_WORKER) */
+				(ticker_id + 1), /* ticker_id */
+				ticks_at_expire, /* current tick */
+				ticks_delay, /* one-shot delayed timeout */
+				0, /* periodic timeout  */
+				0, /* periodic remainder */
+				0, /* lazy, voluntary skips */
+				0,
+				callback, /* handler for executing radio abort or */
+					 /* flash work */
+				context, /* the context for the flash operation */
+				NULL, /* no op callback */
+				NULL);
+
+	return err;
+}
```
Ditto call the implementation in the caller, not necessary to abstract the `ticker_start` call.

```suggestion
```

optical-o replied:
I will keep this one, the function call is in the source code 3 times. It makes it easier to read.

cvinayak reviewed (changes requested):
@jhedberg Lets keep co-ex driver under `drivers/bluetooth/coex` (like co-operation with flash controller is placed in `drivers/flash` , any objections?

@optical-o please add `readme.rst` documentation to reside with implementation, documenting the usage. I believe this is working implementation in this PR meeting your requirements, having documentation to test the feature will help me review/test the PR.

jhedberg commented:
> @jhedberg Lets keep co-ex driver under `drivers/bluetooth/coex` (like co-operation with flash controller is placed in `drivers/flash` , any objections?

I'm not absolutely opposed, but there are a few concerns that come to mind:

The native controller, which I believe is what this primarily concerns, is the only HCI driver that doesn't have any code under `drivers/bluetooth/hci` (rather the code is in `subsys/bluetooth/controller`), so it feels a bit inconsistent to place only this part of controller-related code under `drivers/bluetooth/`

Since this is under `drivers/` I'd expect there to be a clearly defined driver API for it as well, but I don't see that in this PR?

The code is also using private header files from `subsys/bluetooth/controller`, which would speak more in favour of having this code under there too.

cvinayak commented:
> The code is also using private header files from `subsys/bluetooth/controller`, which would speak more in favour of having this code under there too.

ok. We can have the coex driver as `subsys/bluetooth/controller/coex/coex_nrf_ticker.c`. Subsequently move https://github.com/zephyrproject-rtos/zephyr/blob/75bb8ce3826fabe55acb52b18de7116f8a206deb/drivers/flash/soc_flash_nrf_ticker.c into `subsys/bluetooth/controller/flash/flash_nrf_ticker.c`  in a different PR. This way we will have `hci`, `flash` and `coex` which include private header files be in `subsys/bluetooth/controller/{hci,flash,coex}`

optical-o commented:
I will just leave here analyzer screen shot with all of the signals connected. It is an Beacon sample with coex enabled using added prj-coex.conf.
The COEX pin in this case is connected to nRF9160 COEX2 pin configured in "COEX2=4 Mode".

![image](https://user-images.githubusercontent.com/54847502/203755268-2801181e-b1dc-42ea-b679-23f90ef2b18c.png)

Further details for different behaviors of the nRF9160 modem in poor signal conditions can be found [here](https://devzone.nordicsemi.com/f/nordic-q-a/92344/nrf9160-bluetooth-coexistence-insufficient-documentation)

carlescufi commented:
@cvinayak please take a look

cvinayak commented on subsys/bluetooth/controller/coex/coex_ticker.c:
```diff
@@ -0,0 +1,212 @@
+				0,
+				callback, /* handler for executing radio abort or */
+					 /* flash work */
+				context, /* the context for the flash operation */
```
```suggestion
				context, /* the context for the coex operation */
```

cvinayak commented on subsys/bluetooth/controller/coex/coex_ticker.c:
```diff
@@ -0,0 +1,212 @@
+	const struct coex_ticker_config *config = dev->config;
+
+	__ASSERT(ll_radio_state_is_idle(),
+		 "Radio is on during flash operation.\n");
```
```suggestion
		 "Radio is on during coex operation.\n");
```

cvinayak commented on subsys/bluetooth/controller/coex/coex_ticker.c:
```diff
@@ -0,0 +1,212 @@
+				0, /* periodic timeout  */
+				0, /* periodic remainder */
+				0, /* lazy, voluntary skips */
+				0,
```
Why not use reservation in ticker? That is pass a ticks_slot value, this will ensure no other ticker (controller role) expires within this ticks_slot.

ticks_slot would be COEX_RADIO_WORK_RESCHEDULE_DELAY_US, right?


Thinking aloud here, the ticker can be a periodic ticker with  interval and ticks_slot set to the minimum duration coex is required (should I assume this is 200 us? COEX_RADIO_WORK_RESCHEDULE_DELAY_US?). If not, then have it a higher value.
Using ticks_slot ensure there is no redundant BLE radio use during these reserved time that is now owned by coex ticker.

optical-o replied:
I will try to make the modifications based on your feedback.  
Since i don't have deep knowledge of the ticker system inside the controller.  I didn't event know there is this option.

Hopefully i will get to it this week.

cvinayak replied:
The PR is good as-is as a first spin, and I assume that it meets your requirements. You can address the minor change requests  and we should be able to merge this with others reviewers approvals.

In a follow up PR, you can do the improvements that will reduce CPU/Radio use (reduce overall aborts).

cvinayak reviewed (changes requested):
Implementation looks good to me.

Some minor change request, request to add test entry in sample.yaml of the beacon sample so that CI builds/covers the added new feature, and some clarifications from my side.

Adding @mbolivar for DTS and @anangl drivers

cvinayak commented on samples/bluetooth/beacon/sample.yaml:
```diff
@@ -5,3 +5,8 @@ tests:
     harness: bluetooth
     platform_allow: qemu_cortex_m3 qemu_x86 nrf52dk_nrf52832
     tags: bluetooth
+  sample.bluetooth.beacon-coex:
```
We should be good for time being, as you test on nRF52840 DK, we keep just this platform. We add other boards as need be.

cvinayak commented:
You need to rebase...

cvinayak reviewed (dismissed):
Looks good to me on use of Controller's ticker. Others to review...

cvinayak commented:
@optical-o could you address the CI failures?

cvinayak commented on subsys/bluetooth/controller/coex/coex_ticker.c:
```diff
@@ -0,0 +1,212 @@
+#include <errno.h>
+#include <soc.h>
+
+#include <zephyr/zephyr.h>
```
```suggestion
#include <zephyr/kernel.h>
```

cvinayak commented on subsys/bluetooth/controller/coex/readme.rst:
```diff
@@ -0,0 +1,22 @@
+Co-existence Ticker
+###################
+
+Implementation :file:`coex_ticker.c` is designed to utilize co-existence with another transmitter. Chips such as nordic nRF9160 provide a 1-wire co-existence interface, which allows the Bluetooth controller to suspend its activity until the other transceiver suspends its operation.
```
```suggestion
Implementation :file:`coex_ticker.c` is designed to utilize co-existence with another transmitter. Chips such as Nordic Semiconductor nRF9160 provide a 1-wire co-existence interface, which allows the Bluetooth Controller to suspend its activity until the other transceiver suspends its operation.
```

cvinayak commented on subsys/bluetooth/controller/coex/readme.rst:
```diff
@@ -0,0 +1,22 @@
+
+Nordic connect SDK provides detailed description of the 1-wire and 3-wire co-existence interface for the `SoftDevice Bluetooth controller <https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/mpsl/doc/bluetooth_coex.html>`_
+
+Similarly, as in the nordic implementation of the 1-wire interface, the coexistence ticker utilizes a single pin called BLE_GRANT, which active level (high or low) is programmable by the device tree definition.
```
```suggestion
Similarly, as in the nRF connect SDK implementation of the 1-wire interface, in the Zephyr Bluetooth Controller the coexistence ticker utilizes a single pin called BLE_GRANT, wherein active level (high or low) is programmable by the device tree definition.
```

cvinayak commented on subsys/bluetooth/controller/coex/readme.rst:
```diff
@@ -0,0 +1,22 @@
+        inhibit-delay-us = <150>;
+    };
+
+Whenever the grant pin transitions into non-active (such as 1 for the nRF9160). state the implementation starts a ticker job, which in predefined intervals cancels any radio events. This way all advertisements and other radioactivities are suspended until the grant pin transitions into an active state.
```
```suggestion
Whenever the grant pin transitions into non-active (such as 1 for the nRF9160) state the implementation starts a ticker job, which in predefined intervals cancels any radio events. This way all advertisements and other radio events are suspended until the grant pin transitions into an active state.
```

optical-o commented:
@cvinayak I have fixed the CI build after the rebase.
I have included your suggestions. Hopefully this time it will be ready for merge.
Sorry it took so long.

cvinayak reviewed (dismissed):
LGTM on the Controller changes.

gmarull commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,18 @@
+
+  grant-delay-us:
+    type: int
+    default: 150
+    description: |
+        Delay after assertion for the external transceiver operation
```
default needs to be justified, per guidelines. If there's no way to justify it, make it required.

optical-o replied:
Added required to true.

gmarull commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,18 @@
+include: base.yaml
+
+properties:
+  grant-gpios:
+    type: phandle-array
+    description: |
+        GPIO input for the external transceiver
```
missing required?

optical-o replied:
Added required to true.

gmarull commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,18 @@
+description: |
+    Generic representation of Coexistance pinx for radio
+
+compatible: "generic-coex-gpio"
```
how generic is this? from this description I can't understand why.

optical-o replied:
I have updated the description with much longer and hopefully comprehensive one.

gmarull commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,18 @@
+description: |
```
missing license

optical-o replied:
Added licence

cvinayak commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Generic representation of Coexistance pin interafce for radios. This
```
```suggestion
    Generic representation of Coexistance pin interface for radios. This
```

cvinayak commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,31 @@
+    Generic representation of Coexistance pin interafce for radios. This
+    interface is usually available on Wifi/Bluetooth/LTE modules to 
+    interact with each other when sharing same antenna. This prevents
+    any collisions between transmission of different modules. The grant
```
```suggestion
    any collisions between transmissions from different modules. The grant
```

gmarull commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
```diff
@@ -0,0 +1,31 @@
+    transceiver begins their transmission. This is specified by the
+    grant-delay-us property.
+
+compatible: "generic-coex-gpio"
```
I'd probably name this `gpio-radio-coex`, following "gpio-leds" naming,

ref https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#compatible

cvinayak replied:
+1

optical-o replied:
I was following the generic-fem-two-ctrl-pins.yaml example by Nordic.
Should we address the issue or is it blocking for merge ?

optical-o replied:
Changed the compatible to gpio-radio-coex in 2e7ece47945aaa589983aa02c99bfed61744d7d2

cvinayak commented on dts/bindings/net/wireless/generic-coex-gpio.yaml:
why this filename is not `generic,gpio-coex.yaml`? other files in the folder have `Manufacturer,foo-device.yaml` format

gmarull replied:
`generic,` is not a valid manufacturer, and this DT node doesn't actually represent a physical device (if I understood it correctly).

cvinayak replied:
ok, hence based on your earlier proposal, the filename will be `gpio-radio-coex.yaml`, right?

optical-o replied:
The generic-fem-two-ctrl-pins.yaml removed in latest rebase the comma. Not sure why i just followed the example.
The issue might be that the generic vendor has been removed.

optical-o replied:
Renamed the file to in 2e7ece47945aaa589983aa02c99bfed61744d7d2

optical-o committed:
Bluetooth: Controller: Add coexistence implementation

To enable Bluetooth controller coexistence feature, there is
implementation of ticker task, which aborts any ongoing radio events
during assertion of the grant pin. This solves the co-existence issue
in the role of the subordinate transceiver.

Signed-off-by: Tomáš Beneš <tomas@dronetag.cz>

optical-o commented:
I will do final rebase after everything is resolved. For now just to easily track the changes i'm adding commits.

carlescufi commented:
> I will do final rebase after everything is resolved. For now just to easily track the changes i'm adding commits.

@optical-o seems that there's an issue with compliance, can you fix it?

carlescufi commented:
@optical-o 

```
In file included from /__w/zephyr/zephyr/subsys/bluetooth/controller/coex/coex_ticker.c:10:
../../../../../../include/zephyr/zephyr.h:12:2: error: #warning "<zephyr/zephyr.h> is deprecated, include <zephyr/kernel.h> instead" [-Werror=cpp]
   12 | #warning "<zephyr/zephyr.h> is deprecated, include <zephyr/kernel.h> instead"
      |  ^~~~~~~
```

optical-o commented:
Rebased and fixed the include

carlescufi commented:
> Rebased and fixed the include

Thanks, ignore the scancode failure, it's an issue with CI currently.
